File: code/python/03-agent-loop/tools.py
"""Tool implementations and OpenAI function-calling definitions.

Each tool has two parts:
  1. A Python function the orchestrator calls after the LLM requests it.
  2. A JSON schema entry in TOOLS that the LLM sees and uses to decide
     when and how to call the tool.

The docstrings and descriptions are read by the LLM — every word matters.

See docs/02-the-agent-loop/01-anatomy-of-an-agent.md — "The Hands (Tools)"
"""

from __future__ import annotations


# ---------------------------------------------------------------------------
# Tool implementations
# ---------------------------------------------------------------------------


def get_weather(city: str) -> dict:
    """Return current weather conditions for a city.

    In production this would call a weather API (e.g. OpenWeatherMap).
    Mock data is used here so the agent can run without external API keys.

    Args:
        city: City name with optional country code, e.g. "Shanghai, CN".

    Returns:
        dict with keys: city, temperature_c, condition, humidity_percent, wind_kph
    """
    _mock: dict[str, dict] = {
        "Shanghai": {"temperature_c": 22, "condition": "light rain",     "humidity_percent": 85, "wind_kph": 15},
        "London":    {"temperature_c": 14, "condition": "overcast",        "humidity_percent": 78, "wind_kph": 20},
        "New York":  {"temperature_c": 18, "condition": "partly cloudy",   "humidity_percent": 60, "wind_kph": 25},
        "Paris":     {"temperature_c": 16, "condition": "sunny",           "humidity_percent": 55, "wind_kph": 12},
        "Sydney":    {"temperature_c": 28, "condition": "clear",           "humidity_percent": 45, "wind_kph": 18},
    }
    city_key = city.split(",")[0].strip()
    data = _mock.get(city_key, {"temperature_c": 20, "condition": "clear", "humidity_percent": 55, "wind_kph": 10})
    return {"city": city, **data}


def get_stock_price(ticker: str) -> dict:
    """Return current stock price and daily change for a ticker symbol.

    In production this would call a financial data API (e.g. Alpha Vantage).
    Mock data is used here so the agent can run without external API keys.

    Args:
        ticker: Stock ticker symbol in uppercase, e.g. "AAPL".

    Returns:
        dict with keys: ticker, price_usd, change_percent, currency, market_status
    """
    _mock: dict[str, dict] = {
        "AAPL":  {"price_usd": 192.35, "change_percent":  1.2, "currency": "USD"},
        "GOOGL": {"price_usd": 171.80, "change_percent": -0.5, "currency": "USD"},
        "MSFT":  {"price_usd": 415.10, "change_percent":  0.8, "currency": "USD"},
        "TSLA":  {"price_usd": 175.20, "change_percent": -2.3, "currency": "USD"},
        "AMZN":  {"price_usd": 188.40, "change_percent":  0.3, "currency": "USD"},
    }
    ticker_upper = ticker.upper()
    data = _mock.get(ticker_upper, {"price_usd": 100.00, "change_percent": 0.0, "currency": "USD"})
    return {"ticker": ticker_upper, "market_status": "open", **data}


# ---------------------------------------------------------------------------
# OpenAI function-calling definitions
# ---------------------------------------------------------------------------

TOOLS: list[dict] = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": (
                "Get current weather conditions for a city. "
                "Use this when the user asks about weather, temperature, rain, "
                "humidity, wind, or whether to bring an umbrella or coat. "
                "Always call this tool rather than guessing — weather is dynamic."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {
                        "type": "string",
                        "description": (
                            "City name with optional ISO country code, "
                            "e.g. 'Shanghai, CN', 'London, UK', 'New York, US'. "
                            "Include the country code when the city name is ambiguous."
                        ),
                    }
                },
                "required": ["city"],
                "additionalProperties": False,
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "get_stock_price",
            "description": (
                "Get the current stock price and daily percentage change for a publicly "
                "traded company. Use this when the user asks about stock price, share "
                "value, investment potential, or financial performance of a company. "
                "Always call this tool rather than using stale training data."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "ticker": {
                        "type": "string",
                        "description": (
                            "Stock ticker symbol in uppercase, "
                            "e.g. 'AAPL' for Apple, 'GOOGL' for Google, "
                            "'MSFT' for Microsoft, 'TSLA' for Tesla."
                        ),
                    }
                },
                "required": ["ticker"],
                "additionalProperties": False,
            },
        },
    },
]

File: code/nodejs/03-agent-loop/tools.ts
/**
 * Concrete tool implementations: weather, calculator, web search, and date.
 * See: docs/02-the-agent-loop/01-anatomy-of-an-agent.md
 */

import { ToolRegistry, NotFoundError, InvalidArgsError } from "./tool_registry.js";

// ---------------------------------------------------------------------------
// Tool implementations
// ---------------------------------------------------------------------------

const KNOWN_CITIES: Record<string, { temperature: number; condition: string; humidity: number }> = {
  "tokyo, jp": { temperature: 22, condition: "partly cloudy", humidity: 65 },
  "london, uk": { temperature: 14, condition: "overcast", humidity: 80 },
  "new york, us": { temperature: 18, condition: "sunny", humidity: 55 },
  "paris, fr": { temperature: 16, condition: "light rain", humidity: 72 },
  "sydney, au": { temperature: 25, condition: "clear", humidity: 50 },
};

export function getWeather(args: Record<string, unknown>): unknown {
  const city = (args.city as string | undefined)?.toLowerCase().trim() ?? "";
  const weather = KNOWN_CITIES[city];
  if (!weather) {
    const suggestion = Object.keys(KNOWN_CITIES)[0];
    throw new NotFoundError(`City '${city}' not found`, `Try '${suggestion}'`);
  }
  return { city, ...weather };
}

export function calculate(args: Record<string, unknown>): unknown {
  const expr = args.expression as string | undefined;
  if (!expr) throw new InvalidArgsError("expression is required");
  // Safe evaluation: only allow numeric expressions
  const safe = /^[\d\s+\-*/().%]+$/.test(expr);
  if (!safe) throw new InvalidArgsError("Only numeric expressions are allowed");
  try {
    // eslint-disable-next-line no-new-func
    const result = Function(`"use strict"; return (${expr})`)() as number;
    return { expression: expr, result };
  } catch (e) {
    throw new InvalidArgsError(`Could not evaluate: ${String(e)}`);
  }
}

export function webSearch(args: Record<string, unknown>): unknown {
  const query = args.query as string | undefined;
  if (!query) throw new InvalidArgsError("query is required");
  // Mock search results
  return {
    query,
    results: [
      { title: `Result for "${query}"`, url: "https://example.com/1", snippet: `Information about ${query}.` },
      { title: `More about "${query}"`, url: "https://example.com/2", snippet: `Additional details on ${query}.` },
    ],
  };
}

export function getCurrentDate(_args: Record<string, unknown>): unknown {
  const now = new Date();
  return {
    date: now.toISOString().split("T")[0],
    time: now.toTimeString().split(" ")[0],
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    timestamp: now.getTime(),
  };
}

// ---------------------------------------------------------------------------
// Registry builder
// ---------------------------------------------------------------------------

/** Create a ToolRegistry pre-loaded with all demo tools. */
export function createDefaultRegistry(): ToolRegistry {
  const registry = new ToolRegistry();

  registry.register({
    name: "get_weather",
    description: "Get current weather for a city. Returns temperature and conditions.",
    parameters: {
      city: { type: "string", description: "City name with country code, e.g. 'Tokyo, JP'", required: true },
    },
    handler: getWeather,
  });

  registry.register({
    name: "calculate",
    description: "Evaluate a numeric arithmetic expression.",
    parameters: {
      expression: { type: "string", description: "Arithmetic expression, e.g. '2 + 3 * 4'", required: true },
    },
    handler: calculate,
  });

  registry.register({
    name: "web_search",
    description: "Search the web for information about a topic.",
    parameters: {
      query: { type: "string", description: "Search query string", required: true },
    },
    handler: webSearch,
  });

  registry.register({
    name: "get_current_date",
    description: "Get the current date and time.",
    parameters: {},
    handler: getCurrentDate,
  });

  return registry;
}
